Lab7 - Errors

jimbosoft opened the pull request:
Fixes #682

Review of colleague's PR #

#### Changes proposed in this PR: 
Lab 7 - adding error type and added mutex locking to the sync store

## -

Commit:
Added unit test and reduced error checking syntax

Commit:
Added tests for errors

Commit:
Added mutex and chane MapStore into a struct

Commit:
Simplified the error tests and added some for completeness

juliaogris committed:
Remove wllshen8's lab 11 as it leads to failed builds

This is a quick effort to move forward with remaining PRs.

alextmz committed:
Add lab 2, bubblesort (#624)

alextmz committed:
Add lab 6 - CRUD Puppy with interface (#653)

Add lab 6 - CRUD Puppy with interface

alextmz committed:
Add lab 7 - Errors (#655)

Add lab 7 - Errors

alextmz committed:
Add lab 5 - Stringer (#644)

Add lab 5 - Stringer

Commit:
Post lab11 fixes

alextmz committed:
Add lab 4 - Numeronym (#640)

Add lab 4 - Numeronym

alextmz committed:
Add lab 8 - Project Layout (#657)

Add lab 8 - Project Layout

Commit:
Integrated the review points from lab6

1.) Made sure that the items are copied when they are inserted and
removed from the store

2.) Change the generation of the item id to using math.rand in order to
always get a positiv int

3.) Removed the locking from the store that is not the sync store, as
this was not what the instructions asked for

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+package main
+
+func (s *storesSuite) TestReadFailure() {
```
These test functions should be `storer_test.go` along with the other tests. Splitting it up like this just removes the context needed to understand the tests. Also, keeping the successful and failure tests together makes it easier to understand the unit under test from the test cases.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+package main
+
+func (s *storesSuite) TestReadFailure() {
+	pup2, err := s.store.ReadPuppy(1)
```
Why is this `pup2`? Where is `pup` (pup1)? Same applies to the other tests in this file. You use `pup2` when there is no original pup.

This matters because it can be confusing. When I saw `pup2`, I assumed that the test fixture was setting up pup1 for all the tests in the suite, so I went to look at that to see what pup1 was and could not find it. I'm stil not 100% that there is not a pup1 hiding somewhere else (another downside to splitting these tests into a different file - now I don't know how many files I need to look at to find what I'm looking for) and how it may affect these tests.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+	pup2, err := s.store.ReadPuppy(1)
+	s.Require().Nil(pup2)
+	s.Require().Equal(ErrIDNotFound, err)
+	s.Require().NotEmpty(err.Error())
```
This is an odd thing to test. If you want to test that the `ErrIDNotFound` presents as a non-empty string, that test is completely separate from the store test. Furthermore, why is "empty" the useful test for this? If the string was "skdjjfhaksdjafs" - would that be OK?

jimbosoft replied:
Removed

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+package main
+
+func (s *storesSuite) TestReadFailure() {
```
My style for naming tests would be to name this `TestReadIDNotFound`, and the next would be `TestReadValueBelowZero`. The tests share the same prefix (`TestRead`) so I know they are grouped, and the "failure" is specific about exactly which failure is being tested.

This is just my style though, but I think it helps.

jimbosoft replied:
Renamed

camscale commented on 07_errors/jimbotech/storer_test.go:
```diff
@@ -0,0 +1,116 @@
+	id, err := s.store.CreatePuppy(&pup)
+	s.Require().NoError(err)
+	s.Require().NotEqual(pup.ID, uint32(1))
+	s.Require().Equal(id, pup.ID, "Pup id must be set to actual id")
```
This require/assert belongs in `TestCreateSuccess` as that is an invariant of the `CreatePuppy` method. This function is a text fixture called from many places, which means if the storer returned a bad id such that this fails, every test that uses this fixture function will also fail even though what they are testing themselves may well be correct.

jimbosoft replied:
Moved

camscale commented on 07_errors/jimbotech/storer_test.go:
```diff
@@ -0,0 +1,116 @@
+	pup := Puppy{Breed: "kelpie", Colour: brown, Value: "indispensable"}
+	id, err := s.store.CreatePuppy(&pup)
+	s.Require().NoError(err)
+	s.Require().NotEqual(pup.ID, uint32(1))
```
this is non-obvious. Why can't `pup.ID` be `1`? I assume that this fixture function is intended to be called after a test has already created a pup? If so, at the very least this function needs a comment saying so. But I would even say this test is not needed at all - that the storer does not reuse IDs is a test on the `CreatePuppy` function and that test belongs in a test specifically for it.

[time passes as review continues]

oh wow. you generate IDs using `rand`. This makes less sense now. Why can `rand` not return `1`?

jimbosoft replied:
Deleted

camscale commented on 07_errors/jimbotech/storer_test.go:
```diff
@@ -0,0 +1,116 @@
+	s.Equal("kelpie", pup2.Breed)
+	s.Equal(brown, pup2.Colour)
+	s.Equal("indispensable", pup2.Value)
+	s.True(pup2.Colour == brown)
```
Use `S.Equal` here. And all the other places you use `s.True(x == y)`

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/storer_test.go:
```diff
@@ -0,0 +1,116 @@
+	suite.Run(t, &storesSuite{store: &SyncMapStore{}})
+}
+
+//SetupTest creates the correct empty map for each test
```
nitpick: space after double-slash please.

jimbosoft replied:
Fixed

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+
+func (s *storesSuite) TestError() {
+	var err = &Error{}
+	err = nil
```
these two lines can be replaced with `var err *Error` - you specify the type and by not initialising it, you get the zero value, which is nil.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/error_test.go:
```diff
@@ -0,0 +1,45 @@
+	s.Require().Equal(ErrIDNotFound, err)
+}
+
+func (s *storesSuite) TestError() {
```
Should be called `TestErrorNil` or something.

jimbosoft replied:
Renamed

camscale commented on 07_errors/jimbotech/mapstore.go:
```diff
@@ -0,0 +1,74 @@
+	if s == nil {
+		return 0, ErrNotConstructed
+	}
+	p.ID = rand.Int31()
```
wat. i congratulate you on trying to leave land mines for the future developers. I curse you as a possible future developer.

jimbosoft replied:
Previously I used uuid.New().ID() But unfortunately the uint will not turn into a int. This is in my view the next best thing. Similar in functionality. What is your concern? A repeat of the random number? The doco says: "Random numbers are generated by a Source. Top-level functions, such as Float64 and Int, use a default shared Source that produces a deterministic sequence of values each time a program is run." So I expect no repeats during during a run. Maybe I am misunderstanding your objection? What do you suggest?

jimbosoft replied:
Added uniqueness check

camscale commented on 07_errors/jimbotech/mapstore.go:
```diff
@@ -0,0 +1,74 @@
+		return ErrNotConstructed
+	}
+	_, err := s.ReadPuppy(id)
+	if err != nil {
```
Since you do not need the value of the first return value (you're using `_`), idiomatic Go would put this on one line:
```
if _, err := s.ReadPuppy(id); err != nil {
```

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/mapstore.go:
```diff
@@ -0,0 +1,74 @@
+		return ErrNotConstructed
+	}
+	_, err := s.ReadPuppy(id)
+	if err != nil {
```
ditto.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/storer.go:
```diff
@@ -0,0 +1,57 @@
+
+const (
+	// ErrNegativeID error number if id is negative
+	ErrNegativeID = -1
```
These codes should probably be named as `ecNegativeID` or something (ec == error code). Vars starting with `Err` are usually expected to be error types.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+
+// SyncMapStore stores puppies threadsafe.
+type SyncMapStore struct {
+	mux sync.Mutex
```
`mux` is usually the name for a multiplexer, not a mutex. You see that name used a lot for `http.ServeMux`. I think `m` is fine as a name, as it is always used with `Lock()` or `Unlock()` after it - it is clear from usage that `m` is a mutex. Otherwise be verbose and call it `mutex`.

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+	if err != nil {
+		return err
+	}
+	s.mux.Lock()
```
since `sync.Map` is already concurrency safe for individual operations, the only purpose of having a mutex as well is to provide concurrency safety over multiple operations on the map. Yet here you do not take the lock until after you do `ReadPuppy`. That takes and releases the lock, meaning the lock is not held over the duration of all the operations on the `sync.Map`.

The only way this can work then is to not use `ReadPuppy` in this method - you will need to take the lock and directly retrieve from the `sync.Map`

jimbosoft replied:
Not sure what motivated me to do the code this way. Too long ago. The ReadPuppy is called in order to produce the required errors. If I do what you suggest and directly retrieve from the the sync.Map I would have to replicate a lot of code. I guess I put the lock where I did as I wanted to hold it for the shortest possible time, but as you pointed out, another thread can get in between Read and the change. I will move the lock to the top of the function and separate out the error generating retrieval function into unsafeRetrievePuppy.

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+	if err != nil {
+		return err
+	}
+	s.mux.Lock()
```
ditto

jimbosoft replied:
see above

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+func (s *SyncMapStore) CreatePuppy(p *Puppy) (int32, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+	p.ID = rand.Int31()
```
you're evil.

jimbosoft replied:
As above, the gravity of my wickedness is still a mystery to me ... added uniqueness check ...

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+// SyncMapStore stores puppies threadsafe.
+type SyncMapStore struct {
+	mux sync.Mutex
+	sync.Map
```
by embedding the map, you are making the map methods directly accessible to users of `SyncMapStore`, allowing them to bypass any checks/validation you have. You should put the `sync.Map` in as an unexported field.

jimbosoft replied:
Good call.
Done

camscale commented on 07_errors/jimbotech/sync_mapstore.go:
```diff
@@ -0,0 +1,81 @@
+)
+
+// SyncMapStore stores puppies threadsafe.
+type SyncMapStore struct {
```
the filename should be `syncmapstore.go`. But I would call the type `SyncStore` because shorter is better. From the outside, no-one cares that it is a map - it's a store. So don't name it with "Map". Which means the filename could be `syncstore.go`

jimbosoft replied:
Done

camscale commented on 07_errors/jimbotech/storer.go:
```diff
@@ -0,0 +1,57 @@
+
+const (
+	// ErrNegativeID error number if id is negative
+	ErrNegativeID = -1
```
Don't set the error code to -1. Just use `iota` and leave the assignment off subsequent error codes. That let's Go assign the numbers since the actual numbers themselves are not relevant.

jimbosoft replied:
Interesting ... Done

camscale commented on 07_errors/jimbotech/storer.go:
```diff
@@ -0,0 +1,57 @@
+
+// ErrNotConstructed returned if the interface was called without
+// first constructing the underlaying structure.
+var ErrNotConstructed = errors.New("store not created")
```
Why is this not an `Error` like the other ones?

jimbosoft replied:
Because god willed it so ... however if you will something else, that's fine with me too. May the gods argue with each other ... Done

camscale reviewed (commented):
Some comments for you... I'll be back in a few months :)

It would be nice if you could squash your commits into a cohesive set. Leaving old broken commits is like renaming a function to `foo_old()`, fixing it as `foo()` and leaving `foo_old()` in the code.

Also, I edited your PR message to remove the space between `# 682` - that way, github creates it as a link.

Commit:
Made changes as per PR for lab7

Removed error_test and added functions into store_test.go
Renamed sync_mapstore to syncstore.go
Various other renamings and changes

Added syncstore to lab 7

Commit:
Made changes as per PR for lab7

Removed error_test and added functions into store_test.go
Renamed sync_mapstore to syncstore.go
Various other renamings and changes

Added syncstore to lab 7

jimbosoft committed:
Merge branch 'master' into lab7

anzdaddy commented:
Please merge or close.
